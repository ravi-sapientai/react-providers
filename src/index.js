import React from 'react';
import hoistStatics from 'hoist-non-react-statics/dist/hoist-non-react-statics.cjs';

const _contexts = {};

const PENDING = 'pending';
const PROCESSING = 'processing';
const INCLUDED = 'included';

// TODO: use statefull component instead of stateless
export const AppProvider = ({ contexts, children }) => {
  // init context
  Object.keys(contexts).forEach((name) => {
    const { provider, consumer } = contexts[name];
    _contexts[name] = { provider, consumer, status: PENDING };
  });

  const providers = [];

  // create providers graph
  const resolveDependenciesAndAdd = (context) => {
    if (context.status === INCLUDED) {
      return;
    }
    if (context.status === PROCESSING) {
      throw new Error(`Cyclic dependency was detected.`);
    }
    context.status = PROCESSING;

    if (context.provider.contextDependency) {
      context.provider.contextDependency.forEach((contextName) => {
        resolveDependenciesAndAdd(_contexts[contextName]);
      });
    }

    providers.push(context.provider);
    context.status = INCLUDED;
  };

  // Add all providers to graph
  Object.keys(_contexts).forEach((contextName) => {
    resolveDependenciesAndAdd(_contexts[contextName]);
  });

  // render
  return providers.reverse().reduce((store, Provider) => <Provider>{store}</Provider>, children);
};

export const use = (contextName) => (Component) => {
  const fullName = contextName + 'Context';
  if (!_contexts.hasOwnProperty(fullName)) {
    const contextKeys = Object.keys(contexts).toString();
    throw new Error(`There is no such context name ${fullName}.`);
  }

  const { Consumer } = contexts[fullName];

  const Wrap = (parentProps) => (
    <Consumer>
      {(value) => {
        const props = { [contextName]: value };
        return <Component {...parentProps} {...props} />;
      }}
    </Consumer>
  );

  Wrap.displayName = `use(${Component.displayName || Component.name})`;

  if (Component.contextDependency && Array.isArray(Component.contextDependency)) {
    Wrap.contextDependency = [...Component.contextDependency, fullName];
  } else {
    Wrap.contextDependency = [fullName];
  }

  return hoistStatics(Wrap, Component);
};
