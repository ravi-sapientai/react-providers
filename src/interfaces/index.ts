export enum ContextStatuses {
  PENDING = 'pending',
  PROCESSING = 'processing',
  INCLUDED = 'included',
}

export interface IComponentAdditionals {
  contextDependency?: string[];
  displayName?: string;
  name?: string;
}

export interface IContext {
  Provider: React.ComponentType<any> & IComponentAdditionals;
  Consumer: React.ComponentType<any>;
  status?: ContextStatuses;
}

export interface IContextMap {
  [key: string]: IContext;
}
