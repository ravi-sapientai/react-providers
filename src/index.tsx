import { IContextMap, ContextStatuses, IContext, IComponentAdditionals } from './interfaces';
import { use } from './components/use';
import { AppProvider } from './components/AppProvider';

export { AppProvider, use, IContextMap, ContextStatuses, IContext, IComponentAdditionals };
