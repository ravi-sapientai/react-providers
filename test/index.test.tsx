import * as Enzyme from 'enzyme';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { AppProvider, use, IContext } from '../src';

import { contextWithDependency } from './contexts/ContextWithDependency';
import { context, ITodoList } from './contexts/SomeContext';
import { a, b } from './contexts/ContextsWithCyclicDependencies';
import { aContext, bContext, cContext } from './contexts/ComplexDependenciesContexts';

describe('<AppProvider>', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<AppProvider contexts={{}} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = Enzyme.shallow(<AppProvider contexts={{}} />);
    expect(tree).toMatchSnapshot();
  });

  it('adds Context.Provider when rendering', () => {
    const tree = Enzyme.mount(
      <AppProvider contexts={{ some: context }}>
        <div />
      </AppProvider>
    );
    expect(tree).toMatchSnapshot();
  });

  it('resolves dependencies in contexts', () => {
    const tree = Enzyme.mount(
      <AppProvider contexts={{ some: context, contextWithDependency }}>
        <div />
      </AppProvider>
    );
    expect(tree).toMatchSnapshot();
  });

  it('thows an error when dependencies are cyclic', () => {
    const spy = jest.fn();
    class ErrorBoundary extends React.Component {
      state = { error: false };
      componentDidCatch() {
        spy();
      }

      render() {
        return this.props.children;
      }
    }

    const tree = Enzyme.mount(
      <ErrorBoundary>
        <AppProvider contexts={{ a, b }}>
          <div />
        </AppProvider>
      </ErrorBoundary>
    );

    expect(spy.mock.calls.length).toBe(1);
  });

  it('works with complex dependencies e.g. A depends on C, B depends both on A and on C', () => {
    const tree = Enzyme.mount(
      <AppProvider contexts={{ a: aContext, b: bContext, c: cContext }}>
        <div />
      </AppProvider>
    );
    expect(tree).toMatchSnapshot();
  });
});

describe('<Use>', () => {
  class WrappedComponent extends React.Component {
    render() {
      return <div />;
    }
  }
  it('renders without crashing', () => {
    const UseSome = use('some')(WrappedComponent);
    const div = document.createElement('div');
    ReactDOM.render(
      <AppProvider contexts={{ some: context }}>
        <UseSome />
      </AppProvider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders correctly', () => {
    const UseSome = use('some')(WrappedComponent);
    const tree = Enzyme.mount(
      <AppProvider contexts={{ some: context }}>
        <UseSome />
      </AppProvider>
    );

    expect(tree).toMatchSnapshot();
  });

  it('adds access to context data to wrapped component', () => {
    class WrappedComponent extends React.Component<{ some: { data: ITodoList; add: Function } }, {}> {
      componentDidMount() {
        expect(this.props.some.data.list).toEqual([]);
      }
      render() {
        return <div />;
      }
    }
    const UseSome = use('some')(WrappedComponent);
    const tree = Enzyme.mount(
      <AppProvider contexts={{ some: context }}>
        <UseSome />
      </AppProvider>
    );
  });

  it('adds access to context data and methods to wrapped component', () => {
    class WrappedComponent extends React.Component<{ some: { data: ITodoList; add: Function } }, {}> {
      componentDidMount() {
        this.props.some.add('Some todo');
      }
      componentDidUpdate() {
        expect(this.props.some.data.list[0].text).toBe('Some todo');
      }
      render() {
        return <div />;
      }
    }
    const UseSome = use('some')(WrappedComponent);
    const tree = Enzyme.mount(
      <AppProvider contexts={{ some: context }}>
        <UseSome />
      </AppProvider>
    );

    tree.update();
  });
});
