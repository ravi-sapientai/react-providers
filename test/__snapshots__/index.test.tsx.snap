// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppProvider adds Context.Provider when rendering 1`] = `
<Component
  contexts={
    Object {
      "some": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "_defaultValue": Object {},
        },
        "Provider": [Function],
      },
    }
  }
>
  <TodosContextProvider>
    <div />
  </TodosContextProvider>
</Component>
`;

exports[`AppProvider renders correctly 1`] = `""`;

exports[`AppProvider resolves dependencies in contexts 1`] = `
<Component
  contexts={
    Object {
      "contextWithDependency": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "_defaultValue": Object {},
        },
        "Provider": [Function],
      },
      "some": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "_defaultValue": Object {},
        },
        "Provider": [Function],
      },
    }
  }
>
  <TodosContextProvider>
    <use(ContextWithDependencyProvider)>
      <ContextWithDependencyProvider
        some={
          Object {
            "add": [Function],
            "data": Object {
              "list": Array [],
            },
          }
        }
      >
        <div />
      </ContextWithDependencyProvider>
    </use(ContextWithDependencyProvider)>
  </TodosContextProvider>
</Component>
`;

exports[`AppProvider works with complex dependencies e.g. A depends on C, B depends both on A and on C 1`] = `
<Component
  contexts={
    Object {
      "a": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "_defaultValue": Object {},
        },
        "Provider": [Function],
      },
      "b": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "_defaultValue": Object {},
        },
        "Provider": [Function],
      },
      "c": Object {
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {},
          "_currentValue2": Object {},
          "_defaultValue": Object {},
        },
        "Provider": [Function],
      },
    }
  }
>
  <CProvider>
    <use(AProvider)>
      <AProvider
        c={Object {}}
      >
        <use(use(BProvider))>
          <use(BProvider)
            a={
              Object {
                "data": "Hello world",
              }
            }
          >
            <BProvider
              a={
                Object {
                  "data": "Hello world",
                }
              }
              c={Object {}}
            >
              <div />
            </BProvider>
          </use(BProvider)>
        </use(use(BProvider))>
      </AProvider>
    </use(AProvider)>
  </CProvider>
</Component>
`;
